README.md
Useful tools to work with Elastic stack in Python

# Description
    
It consists of two main modules:

- `indextools`: tools to work with Elastic indices
- `doctools`: tools to work with Elastic documents

# Installation
 
## Normal installation

```bash
pip install elastictools
```

## Development installation

```bash
git clone https://github.com/ncthuc/elastictools.git
cd elastictools
pip install --editable .
```

HISTORY.md
# Change log

## [0.2.3] - 2019-06-19
- Fix multiple doc_type in mapping

## [0.2.2] - 2019-06-18
- Fix multiple error related to doc_type in ES7
- Update elasticsearch-py to 7.0.2
- Fix bug related to doctype while indexing document in ES 7


## [0.2.1] - 2019-06-18
- Update elasticsearch-py to 7.0.2
- Fix bug related to doctype while indexing document in ES 7

## [0.1.4] - 2018-12-07

### Added
- `DocTools.dump()`
- `DocTools.bulk_insert_from_json()`

## [0.1.3] - 2018-11-09

### Added
- `DocTools.bulk()`
- `DocTools.bulk_insert_from_csv()`
- `IndexTools.create_if_not_exists()`
- `IndexTools.truncate()`

## [0.1.2] - 2018-11-08

### Added
- `HISTORY.md`
- `long_description` in setup.py that contains content of `README.md` and `HISTORY.md`

### Changed
- `N/A`

### Removed
- `N/A`

## [0.0.1] - 2018-11-07

### Added
- `DocTools.make_search_body()`
- `DocTools.search()`
- `DocTools.msearch()`

### Changed

### Removed

tests/test_doc_tools.py
from elastictools.doctools import DocTools

es = DocTools.from_url('http://localhost:9200')
esi = es.indextool()

tests/test_index_tools.py
from elastictools.indextools import IndexTools

es = IndexTools.from_url('http://localhost:9200')

elastictools/doctools.py
import json
import csv
from collections import deque
import elasticsearch
import elasticsearch.helpers
import jinja2

from elastictools.indextools import IndexTools


class DocTools:
    def __init__(self, hosts=None, es=None):
        """
        Initialize an ElasticSearch instance with list of hosts
        :param hosts: list of host, ex.:
            [
                {'host': 'localhost:9200'},
                {'host': 'othernode', 'port': 443, 'url_prefix': 'es', 'use_ssl': True},
            ]
        """
        self._indextool = None
        if es:
            self._es = es
        else:
            if hosts is None:
                raise ValueError('hosts or es param missing.')
            self._hosts = hosts
            self._es = elasticsearch.Elasticsearch(hosts)

    @classmethod
    def from_url(cls, es_url):
        "Initialize an ElasticSearch with single url"
        hosts = [es_url]
        return cls(hosts=hosts)

    @classmethod
    def from_es(cls, es):
        "Initialize an ElasticSearch instance"
        return cls(es=es)

    def indextool(self):
        """
        Get indextool instance
        :return:
        """
        if not self._indextool:
            self._indextool = IndexTools.from_es(self._es)

        return self._indextool

    @staticmethod
    def render(obj, params):
        """
        Render a jinja2 template
        :param obj: string or dict
        :param params: a dictionary of params
        :return:
        """

        if type(obj) is str:
            t = jinja2.Template(obj)
            return t.render(params)
        else:
            obj = json.dumps(obj)
            t = jinja2.Template(obj)
            return json.loads(t.render(**params))

    def count(self, index_name, body, params, **kwargs):
        """
        Count the number of document in an index, that match the body search
        :param index_name:
        :param body:
        :param params:
        :param kwargs:
        :return:
        """
        if not self.indextool().exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        if params:
            body = DocTools.render(body, params)
        # print(body)
        return self._es.count(index = index_name, body=body)['count']

    def index(self, index_name, body, params=None, id=None, **kwargs):
        """
        Create or update a document
        :param index_name:
        :param body:
        :param params:
        :param id: if None, will generate, if not None, will replace index if existed
        :param kwargs:
        :return:
        """
        if not self.indextool().exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        if params:
            body = DocTools.render(body, params)
        # print(body)
        # fix for ES 7
        # doctype = IndexTools.mapping_get_doctype(self.indextool().get_mapping(index_name))
        doctype = '_doc'
        if id:
            return self._es.index(index = index_name, body=body, doc_type=doctype, id=id, **kwargs)
        else:
            return self._es.index(index=index_name, body=body, doc_type=doctype, **kwargs)

    def delete(self, index_name, id, **kwargs):
        """
        Delete a document with id = `id` in an index
        :param index_name:
        :param id:
        :param kwargs:
        :return:
        """
        if not self.indextool().exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        # doctype = IndexTools.mapping_get_doctype(self.indextool().get_mapping(index_name))
        doctype = '_doc'
        return self._es.delete(index=index_name, id=id, doc_type=doctype, **kwargs)

    def exists(self, index_name, id, **kwargs):
        """
        Check if a document exists in an index or not
        :param index_name:
        :param id:
        :param kwargs:
        :return: boolean
        """
        if not self.indextool().exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        # doctype = IndexTools.mapping_get_doctype(self.indextool().get_mapping(index_name))
        doctype = '_doc'
        return self._es.exists(index=index_name, id=id, doc_type=doctype, **kwargs)

    def get(self, index_name, id, source=False, **kwargs):
        """
        Get a document in an index by it id
        :param index_name:
        :param id:
        :param source:
        :param kwargs:
        :return:
        """
        if not self.indextool().exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        # doctype = IndexTools.mapping_get_doctype(self.indextool().get_mapping(index_name))
        doctype = '_doc'
        if source:
            return self._es.get_source(index=index_name, id=id, doc_type=doctype, **kwargs)
        else:
            return self._es.get(index=index_name, id=id, doc_type=doctype, **kwargs)

    @staticmethod
    def make_search_body(body=None, params=None, from_=None, size=None, query=None, _source=None, highlight=None,
                         aggs=None, sort=None, script_fields=None, post_filter=None, rescore=None, min_score=None,
                         collapse=None, source_includes=None, source_excludes=None):
        """
        Return a body (Python dict) for search query, based on multiple criteria
        :param body:
        :param params:
        :param from_:
        :param size:
        :param query:
        :param _source:
        :param highlight:
        :param aggs:
        :param sort:
        :param script_fields:
        :param post_filter:
        :param rescore:
        :param min_score:
        :param collapse:
        :param source_includes: list of fields
        :param source_excludes:
        :return:
        """
        if not body:
            body = {}

        if query is None:
            query = {"match_all": {}}
        body['query'] = query
        if _source:
            body['_source'] = _source
        if source_excludes or source_includes:
            if not _source:
                _source = {}
            if source_includes:
                _source['includes'] = source_includes
            if source_excludes:
                _source['excludes'] = source_excludes
            body['_source'] = _source

        if highlight:
            body['highlight'] = highlight
        if aggs:
            body['aggs'] = aggs
        if from_:
            body['from'] = from_
        if size:
            body['size'] = size
        if sort:
            body['sort'] = sort
        if script_fields:
            body['script_fields'] = script_fields
        if post_filter:
            body['post_filter'] = post_filter
        if rescore:
            body['rescore'] = rescore
        if min_score:
            body['min_score'] = min_score
        if collapse:
            body['collapse'] = collapse

        if params:
            body = DocTools.render(body, params)

        return body

    def search(self, index_name, body=None, params=None, source_only=False, reserve_id_score=False, **kwargs):
        """
        Execute a search query
        :param index_name:
        :param body:
        :param params:
        :param source_only: get source documents only as Python list, with elastics `_id` and `_score`
        :param kwargs:
        :return:
        """
        if not self.indextool().exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        if params:
            body = DocTools.render(body, params)
        res = self._es.search(index_name, body=body, **kwargs)
        if source_only:
            tmp = res['hits']['hits']
            res = []
            for doc in tmp:
                if reserve_id_score:
                    doc['_source']['_id'] = doc['_id']
                    doc['_source']['_score'] = doc['_score']
                res.append(doc['_source'])
        return res

    def dump(self, index_name, query=None, params=None,
             datetime_field=None, datetime_from=None, datetime_to=None, to_file=False, page_size=1000,
             source_excludes=None, source_includes=None, **kwargs):
        """

        :param index_name:
        :param query:
        :param params:
        :param datetime_field:
        :param datetime_from:  20181101T000000+07:00
        :param datetime_to:    20181107T235959+07:00
        :param to_file:
        :param kwargs:
        :return:
        """
        if not self.indextool().exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))

        sort = None
        if datetime_field:
            sort = [
                {
                  datetime_field: {
                    "order": "asc"
                  }
                }
            ]
            query={
                "bool": {
                    "must": [query] if query else [{"match_all": {}}],
                    "filter": [
                        {
                            "range": {
                                "request_time": {
                                    "gte": datetime_from,
                                    "lt": datetime_to,
                                    "format": "basic_date_time_no_millis"
                                }
                            }
                        }
                    ]
                }
            }
        body = self.make_search_body(query=query, params=params, sort=sort)
        res = self.search(index_name, body=body, source_only=False, **kwargs)
        total = res['hits']['total']
        print('Total: {}'.format(total))
        _from = 0
        _size = page_size
        res = []
        if to_file:
            file = open(to_file, 'w')
            file.write('[')
        while _from < total:
            print('reading {} to {}...'.format(_from+1, min(_from+_size, total)))
            body = self.make_search_body(query=query, params=params, sort=sort, from_=_from, size=_size,
                                         source_includes=source_includes, source_excludes=source_excludes)
            r = self.search(index_name, body=body, source_only=True, **kwargs)
            _from += _size
            if to_file:
                file.write(',\n'.join([json.dumps(rec) for rec in r]) + (',' if _from<total else ''))
            else:
                res.extend(r)
            # res.append(body)

        if to_file:
            file.write(']')
            file.close()
            return total
        return res

    def msearch(self, indices, queries, return_body_only=False, **kwargs):
        """
        Execute a msearch query
        :param return_body_only: if set, not execute the actual search, just return body
        :param indices: list of indices
        :param queries: list of query body
        :param kwargs:
        :return:
        """
        body=''
        for i in range(len(indices)):
            index = json.dumps({'index':indices[i]})
            query = json.dumps(queries[i])
            body += index + '\n' + query + '\n'
        if return_body_only:
            return body
        return self._es.msearch(body=body)

    def bulk(self, index_name, actions, doctype=None, thread_count=1, check_index_existed=True, **kwargs):
        """
        Do bulk actions, if thread_count = 1, otherwise call parallel_bulk
        :param index_name:
        :param actions: any iterable, can also be a generator, in search result format (with `_source`) or orignal format
        :param thread_count: 1 if using bulk, other wise, usi aarop
        :param kwargs:
        :return:
        """
        if check_index_existed:
            if not self.indextool().exists(index_name):
                raise ValueError('index not existed: {}'.format(index_name))

        if not doctype:
            # fix for ES 7
            # doctype = IndexTools.mapping_get_doctype(self.indextool().get_mapping(index_name))
            doctype = '_doc'

        if thread_count<=1:
            print('Normal bulk')
            return elasticsearch.helpers.bulk(self._es, actions, index=index_name, doc_type=doctype, **kwargs)
        else:
            print('Parallel bulk', thread_count)
            return deque(elasticsearch.helpers.parallel_bulk(self._es, actions, index=index_name, doc_type=doctype,
                                                       thread_count=thread_count, **kwargs), maxlen=0)

    def bulk_insert_from_csv(self, filename, index_name, csv_fields=None, thread_count=1, **kwargs):
        """
        bulk insert form csv file
        :param filename:
        :param index_name:
        :param csv_fields: None - use first row as header
        :param thread_count:
        :param kwargs:
        :return:
        """
        with open(filename) as f:
            reader = csv.DictReader(f, fieldnames=csv_fields)
            return self.bulk(index_name, reader, thread_count, **kwargs)

    def bulk_insert_from_json(self, filename, index_name, thread_count=1, **kwargs):
        """
        bulk insert form csv file
        :param filename:
        :param index_name:
        :param thread_count:
        :param kwargs:
        :return:
        """
        with open(filename) as f:
            data = json.load(f)
            return self.bulk(index_name, data, thread_count, **kwargs)

elastictools/indextools.py
import json

import elasticsearch


class IndexTools:
    def __init__(self, hosts=None, es=None):
        """
        Initialize an ElasticSearch instance with list of hosts
        :param hosts: list of host, ex.:
            [
                {'host': 'localhost:9200'},
                {'host': 'othernode', 'port': 443, 'url_prefix': 'es', 'use_ssl': True},
            ]
        """
        self._doctool = None
        if es:
            self._es = es
        else:
            if hosts is None:
                raise ValueError('hosts or es param missing.')
            self._hosts = hosts
            self._es = elasticsearch.Elasticsearch(hosts)

    @classmethod
    def from_url(cls, es_url):
        "Initialize an ElasticSearch with single url"
        hosts = [es_url]
        return cls(hosts=hosts)

    @classmethod
    def from_es(cls, es):
        "Initialize an ElasticSearch instance"
        return cls(es=es)

    # def doctool(self):
    #     if not self._doctool:
    #         self._doctool = DocTool.from_es(self._es)
    #
    #     return self._doctool

    def exists(self, index_name, **kwargs):
        """
        Check if an index or multiple indices existed in ES
        :param index_name: an index name, or list for index names
        :param kwargs:
        :return: True if every index in index_name exists
        """
        return self._es.indices.exists(index_name, **kwargs)

    def exists_type(self, index_name, doc_type, **kwargs):
        """
        Check if a doctytpe existed in a index ES
        :param index_name: an index name, or list for index names, '_all' for all
        :param doc_type: an doc_type name, or list for doc_type names
        :param kwargs:
        :return: True/False
        """
        return self._es.indices.exists_type(index_name, doc_type, **kwargs)

    def get_info(self, index_name, **kwargs):
        """
        Get info of an index
        :param index_name: an index name
        :param kwargs:
        :return:
        """
        if not self.exists(index_name):
            return None

        return self._es.indices.get(index_name, **kwargs)[index_name]

    def get_mapping(self, index_name, **kwargs):
        """
        Get mapping of an index
        :param index_name: an index name
        :param kwargs:
        :return:
        """
        if not self.exists(index_name):
            return None
        return self._es.indices.get_mapping(index=index_name, **kwargs)[index_name]['mappings']

    def clone_mapping(self, index_name, doc_type=None, **kwargs):
        """
        Get mapping of an index
        :param doc_type: new doc type for result mapping, None if unchange
        :param index_name: an index name
        :param kwargs:
        :return:
        """
        if not self.exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        mapping = self._es.indices.get_mapping(index=index_name, **kwargs)[index_name]['mappings']
        if doc_type:
            IndexTools.mapping_set_doctype(mapping, doc_type)
        return mapping

    @staticmethod
    def mapping_get_doctype(mapping):
        """
        Get doctype of the mapping
        :param mapping:
        :return:
        """
        # if len(mapping.keys()) != 1:
        #     raise ValueError('There should be exactly one doc_type in a mapping.')
        # print(mapping.keys())
        key = list(mapping.keys())[0]
        return key

    @staticmethod
    def mapping_set_doctype(mapping, doc_type):
        """
        Set/change doctype of a mapping
        :param mapping:
        :param doc_type:
        :return:
        """
        key = IndexTools.mapping_get_doctype(mapping)
        if key == doc_type:
            return mapping
        mapping[doc_type] = mapping[key]
        mapping.pop(key)
        return mapping

    @staticmethod
    def mapping_cast(mapping, property_name, new_type):
        """
        Change type of a property inside a mapping
        :param mapping:
        :param property_name:
        :param new_type:
        :return:
        """
        if len(mapping.keys()) != 1:
            raise ValueError('There should be exactly one doc_type in a mapping.')
        key = list(mapping.keys())[0]
        if property_name not in mapping[key]:
            return mapping
        if type(new_type) is dict:
            mapping[key][property_name] = new_type
        else:
            mapping[key][property_name]['type'] = new_type
        return mapping

    @staticmethod
    def mapping_rename(mapping, property_name, new_name):
        """
        Rename a property inside a mapping
        :param mapping:
        :param property_name:
        :param new_name:
        :return:
        """
        if len(mapping.keys()) != 1:
            raise ValueError('There should be exactly one doc_type in a mapping.')
        key = list(mapping.keys())[0]
        if property_name not in mapping[key]:
            return mapping

        mapping[key][new_name] = mapping[key][property_name]
        mapping[key].pop(property_name)
        return mapping

    def get_settings(self, index_name, **kwargs):
        """
        Get settings of an index
        :param index_name: an index name, or list for index names, '_all' for all
        :param kwargs:
        :return:
        """
        if not self.exists(index_name):
            return None
        settings = self._es.indices.get_settings(index=index_name, **kwargs)[index_name]['settings']
        return settings

    def clone_settings(self, index_name, **kwargs):
        """
        Clone settings of an index, return dictionary with current index specific data removed
        :param index_name: an index name, or list for index names, '_all' for all
        :param kwargs:
        :return:
        """
        if not self.exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        settings = self._es.indices.get_settings(index=index_name, **kwargs)[index_name]['settings']
        settings['index'].pop('creation_date', None)
        settings['index'].pop('version', None)
        settings['index'].pop('uuid', None)
        settings['index'].pop('provided_name', None)
        return settings

    def stats(self, index_name, **kwargs):
        """
        Get settings of an index
        :param index_name: an index name, or list for index names, '_all' for all
        :param kwargs:
        :return:
        """
        if not self.exists(index_name):
            return None
        return self._es.indices.stats(index=index_name, **kwargs)['indices'][index_name]

    def create(self, index_name, body=None, mapping=None, settings=None, overwrite=False, **kwargs):
        """
        Get settings of an index
        :param body: if specified, ignore settings and mapping
        :param settings:
        :param mapping:
        :param overwrite:
        :param index_name: an index name, or list for index names, '_all' for all
        :param kwargs:
        :return:
        """
        if self.exists(index_name):
            if not overwrite:
                raise ValueError('{} index already existed.'.format(index_name))
            self.delete(index_name)
        if body:
            return self._es.indices.create(index=index_name, body=body, **kwargs)
        else:
            return self._es.indices.create(index=index_name, body={'settings': settings, 'mappings': mapping}, **kwargs)

    def create_if_not_exists(self, index_name, body=None, mapping=None, settings=None, **kwargs):
        if self.exists(index_name):
            return True
        return self.create(index_name, body, mapping, settings, **kwargs)

    def delete(self, index_name, **kwargs):
        """
        Delete an index
        :param index_name:
        :param kwargs:
        :return:
        """
        return self._es.indices.delete(index=index_name, ignore=404, **kwargs)

    def clone(self, src_index, dest_index, mapping=None, settings=None, size=None, script=None, overwrite=None,
              wait_for_completion=False, remote_host=None,
              **kwargs):
        """
        Create dest_index with mapping and settings and reindex src_index into dest_index
        :param src_index: source index name
        :param dest_index: destination index name
        :param mapping: mapping of new index, if None will clone mapping from src_index
        :param settings: settings of new index, if None will clone settings from src_index
        :param kwargs:
        :return:
        """

        remote_es = None

        if not remote_host:
            if not self.exists(src_index):
                raise ValueError('src_index not existed: {}'.format(src_index))
        else:
            remote_es = IndexTools.from_url(remote_host)
            if not remote_es.exists(src_index):
                raise ValueError('src_index {} not existed in remote host {}'.format(src_index, remote_host))

        if not mapping:
            if remote_host:
                mapping = remote_es.clone_mapping(src_index)
            else:
                mapping = self.clone_mapping(src_index)

        if not settings:
            if remote_host:
                settings = remote_es.clone_settings(src_index)
            else:
                settings = self.clone_settings(src_index)
        print('settings', json.dumps(settings))
        print('mapping', json.dumps(mapping))

        self.create(dest_index, mapping=mapping, settings=settings, overwrite=overwrite)

        body = {
            "source": {
                "index": src_index
            },
            "dest": {
                "index": dest_index
            }
        }
        if remote_host:
            body['source']['remote'] = {
                "host": remote_host
            }

        if size:
            body['size'] = size

        if script:
            body['script'] = script

        print("Body:", body)

        return self._es.reindex(body=body, wait_for_completion=wait_for_completion, **kwargs)

    def close(self, index_name, **kwargs):
        """
        close an index
        :param index_name:
        :param kwargs:
        :return:
        """
        if not self.exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        return self._es.indices.close(index=index_name, **kwargs)

    def open(self, index_name, **kwargs):
        """
        Open an index
        :param index_name:
        :param kwargs:
        :return:
        """
        if not self.exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        return self._es.indices.open(index=index_name, **kwargs)

    def reopen(self, index_name, **kwargs):
        """
        this will reload synonym file also
        :param index_name:
        :param kwargs:
        :return:
        """
        if not self.exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        self._es.indices.close(index=index_name, **kwargs)
        return self._es.indices.open(index=index_name, **kwargs)

    def refresh(self, index_name, **kwargs):
        """
        Refresh to make all create/update effected
        :param index_name:
        :param kwargs:
        :return:
        """
        if not self.exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        return self._es.indices.refresh(index=index_name, **kwargs)

    def truncate(self, index_name, wait_for_completion=False, **kwargs):
        """
        Remove all documents in an index
        :param index_name:
        :param wait_for_completion:
        :param kwargs:
        :return:
        """
        if not self.exists(index_name):
            raise ValueError('index not existed: {}'.format(index_name))
        query = {
            "query": {
                "match_all": {}
            }
        }
        return self._es.delete_by_query(index=index_name, body=query, wait_for_completion=wait_for_completion, **kwargs)

    def exists_template(self, template_name, **kwargs):
        """
        Check if an template or multiple templates (comma separated names) existed in ES
        :param template_name: an template_name name, or list for template names
        :param kwargs:
        :return: True if every template in template_name exists
        """
        return self._es.indices.exists_template(template_name, **kwargs)

    def delete_template(self, template_name, **kwargs):
        """
        Delete an template
        :param template_name:
        :param kwargs:
        :return:
        """
        return self._es.indices.delete_template(name=template_name, ignore=404, **kwargs)

    def get_template(self, template_name, **kwargs):
        """
        Get info of an template
        :param template_name: an index template_name
        :param kwargs:
        :return:
        """
        if not self.exists_template(template_name):
            return None

        return self._es.indices.get_template(template_name, **kwargs)[template_name]

    def create_template(self, template_name, patterns, body=None, mapping=None, settings=None, overwrite=False,
                        **kwargs):
        """
        create a template
        :param body: if specified, ignore settings and mapping
        :param settings:
        :param mapping:
        :param overwrite:
        :param template_name: a template name
        :param pattern: template pattern, in array, ex: ["te*", "bar*"]
        :param kwargs:
        :return:
        """
        if not template_name or not patterns:
            raise ValueError('Both template and pattern are required.')
        if self.exists_template(template_name):
            if not overwrite:
                raise ValueError('{} template already existed.'.format(template_name))
            self.delete_template(template_name)
        if body:
            return self._es.indices.put_template(name=template_name, body=body, **kwargs)
        else:
            body = {'index_patterns': patterns,
                    'settings': settings,
                    'mappings': mapping}

            print(json.dumps(body))

            return self._es.indices.put_template(name=template_name, body=body, **kwargs)
